### ISSUE TITLE
NullReferenceException in MvxPathSourceStep.ClearPathSourceBinding

### ISSUE THREAD

**Issue by softlion** [MvvmCross/MvvmCross#1414]
I've stumbled upon an interesting crash today. The crash occured in ClearPathSourceBinding. This is the first time it crashes, in 3 months usage of the app.

```
    private void ClearPathSourceBinding()
    {
        if (this._sourceBinding != null)
        {
            this._sourceBinding.Changed -= this.SourceBindingOnChanged;
            this._sourceBinding.Dispose();
            this._sourceBinding = null;
        }
    }
```

So the only way this can crash, is because it is executed twice in 2 concurrent tasks.
The stacktrace was:

```
MvxPathSourceStep.OnDataContextChanged
MvxSourceStep.DataContext (setter)
MvxFullBinding.DataContext (setter)
MvxTaskBasedBindingContext.OnDataContextChange
Task.Execute (no information from where it has been created)
```

This information is not enough to find the root cause of this problem: i don't know which binding is concerned.

In OnDataContextChange, the DataContext setter is not called from a Task. It is called directly from the DataContext setter. But inside OnDataContextChange are called DataContext setters from bindings.
This should mean it is called from inside the Task of OnDataContextChange.

This means a lock should be added around the code inside MvxPathSourceStep.OnDataContextChanged

```
   private object syncLock = new object();

    protected override void OnDataContextChanged()
    {
        lock(syncLock) {
            this.ClearPathSourceBinding();
            this._sourceBinding = this.SourceBindingFactory.CreateBinding(this.DataContext, this.Description.SourcePropertyPath);
            if (this._sourceBinding != null)
            {
                this._sourceBinding.Changed += this.SourceBindingOnChanged;
            }
        }
        base.OnDataContextChanged();
    }
```

**Comment by kjeremy** [comment]
Oh interesting! I think this might be related to the recyclerview crash.
I'll take a look.

**Comment by alexsorokoletov** [comment]
It's kind of a duplicate of #1390

**Comment by kjeremy** [comment]
I don't quite understand how you're getting into `ClearPathSourceBinding()` multiple times for the same object.  Do we need to start passing `CancellationToken`s around if we're in the middle of altering a `DataContext` to cancel all subsequent tasks?  Where does that leave child bindings?

**Comment by alexsorokoletov** [comment]
@kjeremy If I only knew. We have these problems on iOS, have no idea what's the reason. I know for sure that these problems are not new, since we have fixes in MvvmCross already related to similar crashes just other binding types.

**Comment by kentpickard** [comment]
We recently upgraded to 4.4.0 from 3.5.1 (we stayed there for a long time) and are now seeing these show up as crash reports in HockeyApp (we have 4 crashes so far).  

I am able to reproduce it, but not in a consistent way.  Basically I go to one of our more complicated list views in the iOS app and start scrolling and toggling a checkbox we have in those rows.  After a while of scrolling up and down in a long list, the app will crash.  

Since the task that runs is not awaited, the exception is not raised until the task gets garbage collected which is not predictable.  I would guess that a cell gets re-used so quickly that that there's a race condition.

@kjeremy @alexsorokoletov did you guys ever find a solution for this problem?  I also think this should be marked as a Bug (if you guys are able to that).  

Here's the log from one of the hockey app reports:

```
Incident Identifier: C9ABC7C5-409E-4BA5-8299-FBD09030B348
CrashReporter Key:   A4804E62-C51A-47BB-A2E2-19ECADFB621F
Hardware Model:      iPad4,8
Process:         ...
Path:            /var/containers/Bundle/Application/ABF17062-6B83-434A-8A63-0A4FC894B975/...
Identifier:      ...
Version:         5.0 (0.1.357)
Code Type:       ARM-64
Parent Process:  ??? [1]

Date/Time:       2017-02-21T20:29:04Z
Launch Time:     2017-02-21T20:26:35Z
OS Version:      iPhone OS 10.1.1 (14B100)
Report Version:  104

Exception Type:  SIGABRT
Exception Codes: #0 at 0x18ecbe014
Crashed Thread:  2

Application Specific Information:
*** Terminating app due to uncaught exception 'System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread.', reason: 'System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at MvvmCross.Binding.Bindings.SourceSteps.MvxPathSourceStep.ClearPathSourceBinding () <0x10187b3e4 + 0x00084> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.Bindings.SourceSteps.MvxPathSourceStep.OnDataContextChanged () <0x10187b2dc + 0x00017> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.Bindings.SourceSteps.MvxSourceStep.set_DataContext (System.Object value) <0x10187b65c + 0x00027> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.Bindings.MvxFullBinding.set_DataContext (System.Object value) <0x101876668 + 0x0005f> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.BindingContext.MvxTaskBasedBindingContext.<OnDataContextChange>b__24_0 () <0x1018af330 + 0x0022b> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at System.Threading.Tasks.Task.InnerInvoke () <0x10067705c + 0x0004f> in <15e850188d9f425bbeae90f0bbc51e17#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at System.Threading.Tasks.Task.Execute () <0x1007739a0 + 0x00043> in <15e850188d9f425bbeae90f0bbc51e17#d7d44cb0f4a65bbfdc541bb03b240418>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at MvvmCross.Binding.Bindings.SourceSteps.MvxPathSourceStep.ClearPathSourceBinding () <0x10187b3e4 + 0x00084> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.Bindings.SourceSteps.MvxPathSourceStep.OnDataContextChanged () <0x10187b2dc + 0x00017> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.Bindings.SourceSteps.MvxSourceStep.set_DataContext (System.Object value) <0x10187b65c + 0x00027> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.Bindings.MvxFullBinding.set_DataContext (System.Object value) <0x101876668 + 0x0005f> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at MvvmCross.Binding.BindingContext.MvxTaskBasedBindingContext.<OnDataContextChange>b__24_0 () <0x1018af330 + 0x0022b> in <f9aae2b329aa4d8aa76dab4fe82e828d#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at System.Threading.Tasks.Task.InnerInvoke () <0x10067705c + 0x0004f> in <15e850188d9f425bbeae90f0bbc51e17#d7d44cb0f4a65bbfdc541bb03b240418>:0 
  at System.Threading.Tasks.Task.Execute () <0x1007739a0 + 0x00043> in <15e850188d9f425bbeae90f0bbc51e17#d7d44cb0f4a65bbfdc541bb03b240418>:0 <---


```

**Comment by kentpickard** [comment]
@softlion should have given you a shout out since you originally reported this.  Did you ever find a resolution?  (pls see above comment)

**Comment by alexsorokoletov** [comment]
@kentpickard in my case we have a custom layout structure to avoid using storyboards and to allow better reuse of UITableViewCell designs across app(s).
Because of that we have in the app one place where we call `cell.ClearAllBindings()`. This is where we had the problem. 
We could not wait to re-code the inner MvvmCross for the release and had to try-catch this place.

Might not be ideal solution but it works for us.

**Comment by softlion** [comment]
use a custom base cell and override prepareforreuse to cancel your running tasks. It is always a bad idea to have non cancellable tasks in a cell binding ....

**Comment by kentpickard** [comment]
@alexsorokoletov  I think your earlier suggestion of changing the implementation of `MvxPathSourceStep.ClearPathSourceBinding()` to the following would prevent the null reference exception, but would we be trading a crash for a potential source of memory leaks when the race condition happens and instance assigned to _currentChildBinding is not disposed?
[current implementation](https://github.com/MvvmCross/MvvmCross/blob/8a824c797747f74716fc64c2fd0e8765c29b16ab/MvvmCross/Core/Binding/Bindings/SourceSteps/MvxPathSourceStep.cs#L61)

```
var currentChildBinding = this._currentChildBinding;
this._currentChildBinding = null;
if (currentChildBinding != null)
{
    currentChildBinding.Changed -= this.ChildSourceBindingChanged;
    currentChildBinding.Dispose();
}

```

@softlion I agree that tasks seem problematic in bindings.  However, I don't see a way to cancel the pending tasks that MvvmCross spawns in the `MvxTaskBasedBindingContext.OnDataContextChange()` method [see current source code](https://github.com/MvvmCross/MvvmCross/blob/823efc43aea27ce2d09ab85a8a48014362bef370/MvvmCross/Core/Binding/BindingContext/MvxTaskBasedBindingContext.cs#L152)

What else could I do here in my MvxTableViewCell sub-class?
```
public override void PrepareForReuse()
{
	var context = BindingContext as MvxTaskBasedBindingContext;
	if (context != null)
	{
		// Does not compile
		//context.CancelPendingTasks();

		// How to stop the race condition??

	}
	base.PrepareForReuse();
}
```

At this point I see a few approaches to solving this problem and I'd greatly appreciate your insights:

1. There is a `MvxTaskBasedBindingContext.RunSynchronously` property that can be set to not use tasks.  Seems like it might prevent the race condition (I will test this).  What are your thoughts on the impact of setting RunSynchronously to true?
2. I see in a comment here made by @softlion in  #1423 

> Note that you can easily switch to the legacy binding context, see the instructions there:
> https://mvvmcross.com/v3.2/blog/mvvmcross-42

That link is broken could you re-link?  Is `MvxBindingContext` the legacy binding context [link to source code](https://github.com/MvvmCross/MvvmCross/blob/develop/MvvmCross/Core/Binding/BindingContext/MvxBindingContext.cs)

3. Implement the upgrade proposed by @alexsorokoletov  -- Is there anyway I can replace the implementation of `MvxPathSourceStep` at run-time via IOC registration without Forking the whole repo? 
4. Refactor all my lists to not use bindings (this is the least desirable solution)

What do you guys think?  I do believe this is a bug caused by a race condition that will likely affect apps with complex cells (especially iPad apps).  Is there anyway to get this issue classified as a bug to get more attention?

**Comment by alexsorokoletov** [comment]
@kentpickard well, if my proposed implementation might leak something then we have to make it not leak anything.
What I know for sure that there were few fixes done already regarding the same problem - race condition in bindings. And it was done using locks in other places. I don't like this approach going forward, I believe we can have lock-free code there.

Item 4 you suggest sounds too unreasonable for me :)

**Comment by softlion** [comment]
> I don't see a way to cancel the pending tasks that MvvmCross spawns

You are right. The right way to fix this should be to be able to cancel this task before ClearAllBindings ?

> Proposition 3: you can already go back to the legacy MvxBindingContext. There is a sample or doc somewhere on how to do this easily.

Override RegisterCore

    protected override void RegisterCore()
        {
            Mvx.RegisterSingleton<IMvxBinder>(new MvxFromTextBinder());

            //To get the old behavior back, you can override this registration with
            Mvx.RegisterType<IMvxBindingContext, MvxBindingContext>();
            //Mvx.RegisterType<IMvxBindingContext, MvxTaskBasedBindingContext>();
        }

**Comment by kentpickard** [comment]
For whoever might view this bug in the future, here's how I solved it (still testing but so far so good):

In my sub-class of BaseMvxTableViewSource I override GetCell and then turn off Async behavior by doing this:

// Turn of async binding updates across the board as this was causing some race condition crashes
var bindable = cell as IMvxBindable;
if (bindable != null)
{
	var bindingContext = bindable.BindingContext as MvxTaskBasedBindingContext;
	if (bindingContext != null)
		bindingContext.RunSynchronously = true;
}

I got this idea from [this code in MvxBaseTableViewSource](https://github.com/MvvmCross/MvvmCross/blob/786ca2333a8e770e99e2369938c4aee8d62c5491/MvvmCross/Binding/tvOS/Views/MvxBaseTableViewSource.cs#L109)

which sets RunSynchronously to true if the cells are dynamically sized.  I think MvxBaseTableViewSource should just always turn off Async binding as I have (just need remove that extra condition from the if).

If the main contributors want that upgrade, I'm happy to open a PR.  Just ping me back.

**Comment by martijn00** [comment]
This is fixed in another PR.

**Comment by alexsorokoletov** [comment]
@martijn00 which one, just to track?

**Comment by cjlotz** [comment]
@martijn00 also interested to have a look at the specific full request if at all possible

**Comment by Nickolas-** [comment]
@martijn00 , can you please link PR to this thread that fixed this bug.

**Comment by martijn00** [comment]
I think it was this one: https://github.com/MvvmCross/MvvmCross/pull/1442

**Comment by alborzmgm** [comment]
I think there is the same problem with MvxChainedSourceBinding. A lock statement is required for in Dispose methods.
